### ISSUE TITLE
regression: thread 'tokio-runtime-worker' panicked at 'could not acquire a SinkHandle that can hold the initial since

### ISSUE THREAD

**Issue by philip-stoev** [MaterializeInc/materialize#21316]
### What version of Materialize are you using?

3dc517dfca 

### What is the issue?

Overnight, Nightly on various very-recently-forked-from-main branches has started exhibiting the following panic:

```
default/cluster-u126-replica-133-0[clusterd]: thread 'tokio-runtime-worker' panicked at 'could not acquire a SinkHandle that can hold the initial since Antichain { elements: [1692652498648] }, the since is already at Antichain { elements: [] }', src/storage/src/storage_state.rs:404:13 
```

https://buildkite.com/materialize/nightlies/builds/3421

and 

https://buildkite.com/materialize/nightlies/builds/3414

**Comment by sploiselle** [comment]
tl;dr this is caused by a quick succession of `CREATE` and `DROP SINK`

What appears to happen is:
- Testdrive creates a sink
- The replica on which the sink is scheduled is a little slow to respond
- Testdrive only ensures that the sink is registered with the schema registry, not that is produces data, which I believe occurs outside of and "before" the dataflow
- Testdrive drops the sink:
    - The sink gets removed from the catalog
    - The sink's write frontier gets advanced to `[]`
    - Ostensibly, whatever read hold the sink held is also dropped. I am yet to find the code where the sink takes the read hold, but  can only imagine this happens.
- Testdrive drops the sink's source, which advances its since to `[]`.
- The replica finally starts the sink, but the since of the shard the sink is reading from is `[]`. It panics because it was expecting some time less than `[]`.

To fix this, my suggestion atm is to just make `SinkHandle::new` fallible, and it's allowed to fail iff the error is that it encountered a shard whose since is `[]`. In this case, I think the fact that the read-from shard is `[]` and not just some arbitrarily advanced timestamp is a special case: it indicates that the thing you're reading from was dropped and because of the dependency graph, it means you were also dropped.

I don't think this needs to be a release blocker because the only way for this to occur is rapidly creating and dropping sinks. While we need to fix this, I don't think this kind of timing bug is prevalent enough in the wild to hold back the release. Just my $0.02.

**Comment by petrosagg** [comment]
This code and panic are no more likely in the current version of the code than all the previous ones. If it happened more in nightlies it could be because something in the tests made issuing a `DROP SINK` faster and that's why we hit it more. The behavior that is panicking has been untouched for many many months, so I don't think this should be a release blocker which gives us a bit more time to implement a proper fix.

**Comment by philip-stoev** [comment]
Happened again https://buildkite.com/materialize/nightlies/builds/3545

**Comment by petrosagg** [comment]
>  this fix may also be fairly trivial:

This will be work that would have to be undone when i get to my controller work (because the assertion is correct in that setting) and I fear it might get dropped/forgotten. This issue has been open for a long time so I would leave it be for a few more months.

**Comment by def-** [comment]
Still happening in CI and causing test failures btw: https://buildkite.com/materialize/tests/builds/78668#018e56ba-5138-4b46-8a4b-e4770e4c9234
